File: src/main/java/com/tabnine/inline/render/preserved/InlayElementRenderer.java
package com.tabnine.inline.render.preserved;

import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.EditorCustomElementRenderer;
import com.intellij.openapi.editor.Inlay;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.ui.JBColor;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

import static com.tabnine.inline.render.FontFinder.getFont;

public class InlayElementRenderer implements EditorCustomElementRenderer {
    private final Editor editor;
    private final String suffix;
    private final boolean deprecated;

    public InlayElementRenderer(Editor editor, String suffix, boolean deprecated) {
        this.editor = editor;
        this.suffix = suffix;
        this.deprecated = deprecated;
    }

    @Override
    public int calcWidthInPixels(@NotNull Inlay inlay) {
        return editor.getContentComponent().getFontMetrics(getFont(editor, deprecated)).stringWidth(suffix);
    }

    @Override
    public void paint(
            @NotNull Inlay inlay,
            @NotNull Graphics g,
            @NotNull Rectangle targetRegion,
            @NotNull TextAttributes textAttributes) {
        g.setColor(JBColor.GRAY);
        g.setFont(getFont(editor, deprecated));
        g.drawString(suffix, targetRegion.x, targetRegion.y + editor.getAscent());
    }
}

File: src/main/java/com/tabnine/inline/render/preserved/InlayElementRenderer.kt
package com.tabnine.inline.render.preserved

import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.EditorCustomElementRenderer
import com.intellij.openapi.editor.Inlay
import com.intellij.openapi.editor.markup.TextAttributes
import com.intellij.ui.JBColor
import com.tabnine.inline.render.GraphicsUtils
import java.awt.Graphics
import java.awt.Rectangle

class InlayElementRenderer(private val editor: Editor, private val suffix: String, private val deprecated: Boolean) :
    EditorCustomElementRenderer {
    override fun calcWidthInPixels(inlay: Inlay<*>): Int {
        return editor.contentComponent.getFontMetrics(GraphicsUtils.getFont(editor, deprecated)).stringWidth(suffix)
    }

    override fun paint(
        inlay: Inlay<*>,
        g: Graphics,
        targetRegion: Rectangle,
        textAttributes: TextAttributes
    ) {
        g.color = JBColor.GRAY
        g.font = GraphicsUtils.getFont(editor, deprecated)
        g.drawString(suffix, targetRegion.x, targetRegion.y + editor.ascent)
    }
}
